### DISCUSSION TITLE
NXT as pipeline tool

### DISCUSSION THREAD

**Discussion by petertimberlake** [nxt-dev/nxt_editor#208] [Q&A] [answered]
Hi!

I'm very interested in using NXT as a sort of 'connective tissue' between all of the applications in our pipeline. For example, it would manage a workflow that may look something like:

_kick off maya render > create .mp4 proxy > denoise render with an automatically generated .nk script > update nuke script with new cg render > render nuke script_

This would require a lot of command line instancing of various applications, probably via some kind of subprocess-like thing.

My questions are:
1. Are there established ways of approaching this? Or creative ideas about how this would work?
2. Imagine an "update nuke script with cg render" script is like 50 lines or something. What would be the best way to pass the contents of that node to an "execute bash script" node. I'm hoping for a way that's more elegant than stringifying all python before passing an entire python script as a single stringified variable? 

BONUS QUESTION:
Could anyone point me to NXT implementations with nuke I may look at?

**Comment by ImLucasBrown** [comment, marked as answer]
Hi Peter,

Really great questions. The workflow you're describing is something we see a lot of value in, but it is one of our least developed feature sets. We have a rough framework for running multi-DCC operations in NXT, but the API is very likely to change as we have time to iterate on the problem space.

We call running graphs in other interpreters "contexts", we have some slim documentation about setting up a context [here](https://nxt-dev.github.io/extensions/#plugins) and a video covering how you can run a graph in a context [here](https://nxt-dev.github.io/tutorials/#contexts). 

1. While we don't "officially" support Nuke, I do have some demos of running an interconnected set of graphs, one of which is Nuke. [This video](https://youtu.be/Sli_QXtCoz4) shows standalone Python running, passing data to Maya, and finally to Nuke. I've attached a zip of those graphs here so you can open them up and see how they work. Start with the graph named `MultiContext.nxt`. Assuming you set up your contexts following the doc links above you should be able to run these in current state NXT. You'll need to start NXT as a standalone app (not in a DCC) though as it has an RPC server running in the background that does the magic.

2. If I understand correctly you're hoping to have a "bash script” node. We don't have support for that (yet). However, we don't prevent calls to `subprocess` or `os.system` to run arbitrary commands in the shell. If you're looking to load the contents of a file into a node we have a special token for that: `${contents::}`. You can nest tokens so loading a file is really quite easy, something like this will work `${contents::${file::my_script.py}}`.

This is a problem space we're keen to get more workflows and input on. NXT can manage the spin up of any arbitrary Python interpreter so Nuke and Maya are very easy as they ship their own interpreters. DCCs like Unreal Engine have been harder, but not impossible as shown [here](https://youtu.be/hyg60PwZH5Y), @MichaelAldrich can speak more to that. In simple terms, right now we depend on being able to run `path/to/python.exe -m`.
If you’d like to speak with the whole NXT team and possibly some of our active users, we have a monthly Zoom meeting that’s open to the public. Our next meeting is September 26th, we can get you an invite if you’re interested.
In anycase, if you have more questions we love to talk about NXT!



[multi_context.zip](https://github.com/nxt-dev/nxt_editor/files/7035738/multi_context.zip)

**Reply by petertimberlake** [reply]
Thank you Lucas!

This is super helpful, and thanks so much for your continued development and support with NXT! This seems to be exactly what I'm looking for.

I feel I'm close to getting this to work, but there seems to be a problem with registering the nuke context. Any ideas on what's going wrong? Working with NXT Standalone Editor v3.10 / API v0.12.0 / Python 3.7.9 on Windows  I'm receiving the error:

`xmlrpc.client.Fault: :Unknown context "nuke12_context"`

I've tried placing these files `nuke_context.py` and `nuke12_context.nxt` in the following directories with no success:

`C:\Users\admin\nxt\config\0\plugins`
`C:\NXT\nxt\config\0\plugins`
`C:\NXT\nxt\plugins`
`C:\NXT\nxt\remote`

`nuke_context.py` follows the `uhoh_context` sample code:

```
# Builtin
import os

# External
from nxt.remote.contexts import RemoteContext, register_context

# Setup Context

_name = 'nuke12_context'
_exe = 'C:/Program Files/Nuke12.2v3/Nuke12.2.exe'
_graph = 'C:/Users/admin/nxt/config/0/plugins/nuke12_context.nxt'
_args = ()
_context = RemoteContext(_name, _exe, _graph, _args)
register_context(_context)
```

and `nuke12_context.nxt` 

```
{
    "version": "1.17",
    "alias": "nuke12_context",
    "color": null,
    "mute": false,
    "solo": false,
    "references": [
        "$NXT_BUILTINS/_context.nxt"
    ]
}
```

Also tried one which follows the blender template and swaps out the blender-specific stuff with nuke specific stuff.

Any ideas? Thanks again for your help!

Edit: also would love to attend the next NXT meeting !

**Reply by ImLucasBrown** [reply]
That's an odd error, everything looks fine. Not sure what would cause it to not load plugins at `%UserProfile%/nxt/config/0/plugins`. The only thing I see wrong, which wouldn't cause the `unknown context` error, is you need to set the `_exe` path to the Nuke Python eg `C:\Program Files\Nuke11.1v2\python.exe`.


You could try setting the `NXT_SITE_DIR` env var, basically make a folder structure like this: `C:/SomeDir/config/0/plugins` and drop the `nuke_context.py` and `nuke12_context.nxt` into there. The `NXT_SITE_DIR` is how we've set it up in studio pipelines so multiple machines can access the same context plugin.

---

The Zoom meeting info is at the bottom here.
Our next one is Sep 26 from 11am - 12pm CST (utc-6)
```
Meeting ID: 369 337 4599
Passcode: bTj5XZ
```
